src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs

```diff
@@ -0,0 +1,536 @@
+        /// </summary>
+        private bool HeaderWasNotRequested(ChainedHeader chainedHeader)
+        {
+            return !((chainedHeader.BlockDataAvailability == BlockDataAvailabilityState.BlockAvailable)
+                  || (chainedHeader.BlockDataAvailability == BlockDataAvailabilityState.BlockRequired));
+        }
+
+        /// <summary>
+        /// Check whether a header is in one of the following states
+        /// <see cref="ValidationState.AssumedValid"/>, <see cref="ValidationState.PartiallyValidated"/>, <see cref="ValidationState.FullyValidated"/>.
+        /// </summary>
+        private bool HeaderWasNotMarkedAsValidated(ChainedHeader chainedHeader)
+        {
+            return !((chainedHeader.BlockValidationState == ValidationState.AssumedValid)
+                  || (chainedHeader.BlockValidationState == ValidationState.PartiallyValidated)
+                  || (chainedHeader.BlockValidationState == ValidationState.FullyValidated));
+        }
+
+        private void RemoveUnclaimedBranch(ChainedHeader chainedHeader)
+        {
+            this.logger.LogTrace("({0}:'{1}')", nameof(chainedHeader), chainedHeader);
+
+            ChainedHeader currentHeader = chainedHeader;
+            while (true)
+            {
+                bool headerHasNextHeader = currentHeader.Next.Count > 0;
+                if (headerHasNextHeader)
+                {
+                    this.logger.LogTrace("Header '{0}' is part of another branch.", currentHeader);
+                    break;
+                }
+
+                bool headerHasPeerClaim = this.peerIdsByTipHash.ContainsKey(chainedHeader.HashBlock);
+                if (headerHasPeerClaim)
+                {
+                    this.logger.LogTrace("Header '{0}' is claimed by a peer and wont be removed.", currentHeader);
+                    break;
+                }
+
+                this.logger.LogTrace("Header '{0}' was removed from the chain tree.", currentHeader);
+                this.chainedHeadersByHash.Remove(currentHeader.HashBlock);
+                currentHeader.Previous.Next.Remove(currentHeader);
+
+                currentHeader = currentHeader.Previous;
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>
+        /// Remove the peer's tip and all the headers claimed by this peer unless they are also claimed by other peers.
+        /// </summary>
+        /// <param name="chainedHeader">The header where we start walking back the chain from.</param>
+        /// <param name="networkPeerId">The peer id that is removed.</param>
+        private void RemovePeerClaim(int networkPeerId, ChainedHeader chainedHeader)
+        {
+            this.logger.LogTrace("({0}:{1},{2}:'{3}')", nameof(networkPeerId), networkPeerId, nameof(chainedHeader), chainedHeader);
+
+            var listOfPeersClaimingThisHeader = this.peerIdsByTipHash.TryGet(chainedHeader.HashBlock);
```

Comment:
It's not listOfPeersClaimingThisHeader, it's more like a list of peers that claim this header as their tip. 
It is possible that peer will claim a header which ancestor is the target header.

Reply:
Also unvar